app/lib/exchange-rate/money.ts

```diff
@@ -0,0 +1,480 @@
+    return this._data.amount.lte(amountInMinUnit);
+  };
+
+  isZero = (): boolean => {
+    return this._data.amount.eq(0);
+  };
+
+  isPositive = (): boolean => {
+    return this._data.amount.gt(0);
+  };
+
+  isNegative = (): boolean => {
+    return this._data.amount.lt(0);
+  };
+
+  /**
+   * Converts money to a regular javascript number for the requested currency unit.
+   * If no unit is provided, base/default unit is used (bitcoin for BTC, dollar for USD, etc.)
+   * Throws an error if it's not possible to keep full precision.
+   */
+  toNumber = (unit?: string): number => {
+    // Don't use big.js toNumber because it sometimes returns -0.
+    const str = this.toString(unit);
+    const num = Number(str);
+
+    if (
+      str !==
+      this.merge({
+        amount: num,
+        unit: unit ?? this.currencyData.baseUnit,
+      }).toString(unit)
+    ) {
+      throw new Error('Converting to number was imprecise');
+    }
+
+    return num;
+  };
+
+  /**
+   * Returns the money value as a stringified number without a currency.
+   *
+   * If you want to include the currency as well use `toLocaleString` instead.
+   *
+   * @param unit Specifies the currency unit to return. If not provided the base/default unit is used (bitcoin for BTC,
+   * dollar for USD, etc.)
+   */
+  toString = (unit?: string): string => {
+    const currencyUnit = this.getCurrencyUnit(unit);
+    const amount = this.amount(currencyUnit.name);
+    return amount.toFixed(currencyUnit.decimals);
+  };
+
+  /**
+   * Returns the money value as a localized stringified number.
+   * @param locale Locale to use to format the number. If not provided the locale of the machine will be used.
+   * @param unit Specifies the currency unit to return. If not provided the base/default unit is used (bitcoin for BTC,
+   * dollar for USD, etc.)
+   * @param showCurrency Controls if the currency symbol should be included in the string. True by default
+   */
+  toLocaleString = ({
+    locale,
+    unit,
+    showCurrency = true,
+  }: {
+    locale?: string;
+    unit?: string;
+    showCurrency?: boolean;
+  } = {}): string => {
+    const currencyUnit = this.getCurrencyUnit(unit);
+    const options: FormatOptions = { locale };
+    if (showCurrency) {
+      options.currency = this.currency;
+    }
+    return currencyUnit.format(this.toNumber(unit), options);
+  };
+
+  toJSON = () => {
+    return {
+      amount: this.toNumber(),
+      currency: this.currency,
+    };
+  };
+
+  /**
+   * Converts the money to the provided currency based on the provided exchange rate.
+   * @param currency Currency to convert the money to (target currency)
+   * @param exchangeRate Exchange rate to apply. The rate has to be in source/target currency format. E.g. if converting
+   * USD to BTC, the rate should be in USD/BTC format. If converting BTC to usd it should be in USD/BTC format.
+   */
+  convert = (currency: string, exchangeRate: NumberInput): Money => {
```

Comment:
So in `ExchangeRateService` we will only ever have BTC/USD, not USD/BTC. Should we add support for inverse exchange rates to the `ExchangeRateService`? 

I'm just wondering how we will use this in practice, it seems prone to bugs.

This way we would have to do something like:

```js
const rates: Rates = {
   BTC/USD: 100_000
}

// $1
const usdMoney = Money.create({ amount: 1, currency: "USD", unit: "dollar"});
// amount of btc for 1 dollar
const amountBtc = usdMoney.convert('BTC', 1 / rates[BTC/USD])
```

Or more generally I wonder if there's a more clear way to do this so that we are sure the correct exchange rate is being used rather than leaving it up to the implementation

Reply:
Yeah I would update exchange rate providers to calculate the inverse rates as well. then when calling exchange rate service you can ask for both tickers and just pick the one you need. so you would have something like:
```
const rates: Rates = {
   BTC/USD: 100_000
   USD/BTC: ...
}

// $1
const usdMoney = Money.create({ amount: 1, currency: "USD", unit: "usd"});
// amount of btc for 1 dollar
const amountBtc = usdMoney.convert('BTC', rates[USD/BTC])
```